src/daft-core/src/array/ops/clip.rs

```diff
@@ -0,0 +1,106 @@
+use arrow2::array::PrimitiveArray;
+use common_error::{DaftError, DaftResult};
+use num_traits::{clamp, clamp_max, clamp_min};
+
+use crate::{array::DataArray, datatypes::DaftNumericType, prelude::AsArrow};
+
+/// Helper function to perform clamping on a value with optional bounds using `Option<T>` directly
+fn clamp_helper<T: PartialOrd + Copy>(
+    value: Option<&T>,
+    left_bound: Option<&T>,
+    right_bound: Option<&T>,
+) -> Option<T> {
+    match (value, left_bound, right_bound) {
+        (None, _, _) => None,
+        (Some(v), Some(l), Some(r)) => {
+            assert!(l <= r, "Left bound is greater than right bound");
+            Some(clamp(*v, *l, *r))
+        }
+        (Some(v), Some(l), None) => Some(clamp_min(*v, *l)),
+        (Some(v), None, Some(r)) => Some(clamp_max(*v, *r)),
+        (Some(v), None, None) => Some(*v),
+    }
+}
```

Comment:
They key observation we can leverage here for some better performance is that the result of the clamp is None if the original value is None. Therefore instead of doing `as_arrow.iter()` you can use `as_arrow.values_iter()`, which will return an iterator of all the values, ignoring the validity. This is fine because we slap on the validity of the original array anyway. The very small benefit of this is that it will reduce the number of match branches, i think only by 1 or something.

Unfortunately we can't do this for left and right though, because we need to account for their validity.

But in a case like `(array_size, 1, rbound_size)` and the single left_bound is not None, you only need 1 validity check per row! i.e. for the right_bound (because you are using `values_iter` for the array, and your left bound is a non-null scalar). 

Lastly, and probably the most important, in the case of `(_, 1, 1)` you can probably do something like 
```
let left = left_bound.get(0);
let right = right_bound.get(0);
if let Some(left) = left
    && let Some(right) = right
{
    self.apply(|v| clamp(v, left, right))
} else if let Some(left) = left {
    self.apply(|v| clamp_min(v, left))
} else if let Some(right) = right {
    self.apply(|v| clamp_max(v, right))
} else {
    Ok(Self::full_null(self.name(), self.data_type(), self.len()))
}
```

Reply:
Makes sense.

I've removed the `clamp_helper` functions and re-wrote the function that I pass into the `map` to reduce the amount of unneeded `match` arms for the various cases. Maybe a macro would be good, but I think this is clear enough for now.